### ISSUE TITLE
OpenGL binding model

### ISSUE THREAD

**Issue by kvark** [gfx-rs/gfx#2784]
Related to the implementation in #2071 and general limitations of #1619.

## Problem
GLSL has this super annoying limitation of only having combined image samplers. API itself has seen the addition of separate "sampler object" concept, but the shading language never caught up to it :( What this means for us is basically that this Vulkan GLSL code is not compatible with OpenGL:
```glsl
layout(set = 0, binding = 0) uniform sampler s_PaletteSampler;
layout(set = 1, binding = 0) uniform utexture1D t_ColorTable;
layout(set = 1, binding = 1) uniform texture1D t_Palette;
layout(set = 1, binding = 2) uniform sampler s_ColorTableSampler;
layout(set = 1, binding = 3) uniform sampler s_OtherSampler;
void main() {
    uvec2 color_params = texelFetch(usampler1D(t_ColorTable, s_ColorTableSampler), 5, 0).xy;
    uvec2 other_params = texelFetch(usampler1D(t_ColorTable, s_OtherSampler), 6, 0).xy;
    vec4 main_color = texelFetch(sampler1D(t_Palette, s_PaletteSampler), int(color_params[0]), 0);
    vec4 other_color = texelFetch(sampler1D(t_Palette, s_OtherSampler), int(other_params[0]), 0);
}
```
On the level of the binding model we have 2 images and 2 samplers, but OpenGL-style GLSL would have to use 4 different "sampler" objects (which are combined), and this number can grow as the number of combinations of images used with samplers grows in a program...

Note: this problem applies equally to gfx-rs, Vulkan Portability, and WebGPU.
cc @msiglreith @ZeGentzy @grovesNL @Kangz

## Solution
There is no good way to solve it, but we need a strict rule defined that could be:
  1. efficiently implemented by the backend without diverging too much from the Vulkan binding model (in particular, without introducing too much lazyness in binding and pipeline creation).
  2. clearly communicated to the clients and verified on our side as a part of the portability contract, while still permitting typical use-cases to work unchanged.

My proposal is to have the following rule: each image needs to be used by a shader with *no more than one sampler*, known at pipeline creation time. This means that the binding model would only have a single slot for this image, and that it could persist through the changing pipelines as a command buffer is recorded.

In terms of implementation, when a new pipeline is set, it needs to bind the GL sampler objects to the active texture slots used by the shader, corresponding to that constant association of the images with samplers in the shader (that pipeline knows about).

## Challenges
The idea isn't fully thought through or tested yet. I imagine one of the difficulties would be extracting the image-sampler association from the SPIR-V. Some analysis is needed that would take into account uniform control flow and specialization constants...

Another problem is what do we do about client code that doesn't fit into the rule. It could be useful to provide an ability for the user to say, at the pipeline layout (or descriptor set layout) creation time, that a particular image will be used with more than one sampler, without compromising the general effectiveness of the binding model.

**Comment by Kangz** [comment]
> My proposal is to have the following rule: each image needs to be used by a shader with no more than one sampler, known at pipeline creation time. This means that the binding model would only have a single slot for this image, and that it could persist through the changing pipelines as a command buffer is recorded.

This should be strengthened a little bit because right now it is hard to describe what *known at pipeline compilation time* means. I'd suggest saying that the valid arguments for `OpTypeSampledImage` construction are variable defined in the interface. This way it is trivially easy to know which images to create. There might be more complications for when `OpImage` is used directly though, not sure.

**Comment by munrocket** [comment]
We can just create 1 additional texture for all texture coordinates. For example in this WebGL1 code
```
uniform vec2 texcoord1;
uniform vec2 texcoord2;
uniform vec2 texcoord3;
uniform sampler2D colorSampler1;
uniform sampler2D colorSampler2;
uniform sampler2D colorSampler3;
void main() {
  vec4 color1 = texture2D(colorSampler1, texcoord1);
  vec4 color2 = texture2D(colorSampler2, texcoord2);
  vec4 color3 = texture2D(colorSampler3, texcoord3);
  gl_FragColor = vec4(color1.x, color2.x, color3.x, 1.0);
}
```
We can modify like that
```
unifrom int count;
uniform samler2D coordSampler;                            // Nx1 texture
uniform sampler2D colorSampler[count];
void main() {
  vec4 colors[count];
  for (int i = 0; i < count; i++) {              // loop with non-constant expression
    vec2 texcoord = texture2D(coordSampler, vec2(float(i) / count, 0));
    colors[i] = texture2D(colorSampler[i], coords);        //dynamic array indexing
  }
  gl_FragColor = vec4(colors[0].x, colors[1].x, colors[2].x, 1.0);
}
```
But unfortunately this will work only in WebGL2, because dynamic array index not supported in WebGL2, also we have non-constant expression in loop. So... According to [maximum](https://webglstats.com/webgl/parameter/MAX_VERTEX_TEXTURE_IMAGE_UNITS) number of texture units we can do something like that
```
#define MAXTEX 15
unifrom int count;
uniform sampler2D coordSampler;
uniform sampler2D colorSampler[count];

void main() {
  vec4 colors[count];
  for (int i = 0; i < MAXTEX; i++) {
    if (i == count) break;

    vec2 texcoord = texture2D(coordSampler, vec2(float(i) / count, 0.5));

    if (i == 0) colors[0] = texture2D(colorSampler[0], texcoord);
    if (i == 1) colors[1] = texture2D(colorSampler[1], texcoord);
    ....
    if (i == 14) colors[13] = texture2D(colorSampler[13], texcoord);
    if (i == 15) colors[14] = texture2D(colorSampler[14], texcoord);
  }
  gl_FragColor = vec4(colors[0].x, colors[1].x, colors[2].x, 1.0);
}
```
in other words we can to inline in code some strange function where colorSampler appear in OGL 4.5 shader
```
texture2DHelper(samplerId, texcoord) {
  if (samplerId == 0) texture2D(colorSampler[0], texcoord);
  if (samplerId == 1) texture2D(colorSampler[1], texcoord);
  ....
  if (samplerId == 13) texture2D(colorSampler[13], texcoord);
  if (samplerId == 14) texture2D(colorSampler[14], texcoord);
}
```
@kvark I can't understand your shader example is not compatible. The only thing that I know is that texelFetch is not supported by WebGL1.
